### PR TITLE
feat(roi_cluster_fusion): check whether cluster RoI is inside of RoI detected by 2D detector

### PR DESCRIPTION
**Pull request by ktro2828**
## Description

This PR updates `roi_cluster_fusion` to prevent FP, which detects shrubbery as a car.

This PR adds support of **checking whether cluster RoI is inside of RoI detected by 2D detector**.
I added the following function to achieve the above.

```c++
/**
* @param outer : RoI detected by 2D detector
* @param inner : Cluster RoI
* @param outer_offset_scale : A scale factor to extend outer RoI a little. Defaults to 1.1.
*/
bool is_inside(
  const sensor_msgs::msg::RegionOfInterest & outer,
  const sensor_msgs::msg::RegionOfInterest & inner, const double outer_offset_scale = 1.1);
```

## Related links

[TIER IV INTERNAL LINK](https://tier4.atlassian.net/browse/RT1-2663)

## Tests performed

**Before**

https://github.com/autowarefoundation/autoware.universe/assets/60615504/20eb8ba9-27af-4d64-9e67-d466bf643945

**After**

https://github.com/autowarefoundation/autoware.universe/assets/60615504/ee99cc95-12d1-4965-8784-da9b7442df6e

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp [hunk 1]
```diff
@@ -177,7 +177,7 @@ void RoiClusterFusionNode::fuseOnSingleImage(
       if (use_iou_y_) {
         iou_y = calcIoUY(cluster_map.second, feature_obj.feature.roi);
       }
-      if (max_iou < iou + iou_x + iou_y) {
+      if (max_iou < iou + iou_x + iou_y && is_inside(feature_obj.feature.roi, cluster_map.second)) {
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp:180]: I would like to have the option to choose.

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/utils/geometry.hpp [hunk 2]
```diff
@@ -60,6 +60,10 @@ void transformPoints(
   const std::vector<Eigen::Vector3d> & input_points, const Eigen::Affine3d & affine_transform,
   std::vector<Eigen::Vector3d> & output_points);
 
+bool is_inside(
+  const sensor_msgs::msg::RegionOfInterest & outer,
+  const sensor_msgs::msg::RegionOfInterest & inner, const double outer_offset_scale = 1.1);
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/include/image_projection_based_fusion/utils/geometry.hpp:65]: I would like to change scale factor by launch parameter.

File: perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp (hunk 1 above)
Comment (ktro2828) [reply perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp]: updated in https://github.com/autowarefoundation/autoware.universe/pull/4147/commits/9fa0df5259dc534859c4b3beda72ab09640a1acd

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/utils/geometry.hpp (hunk 2 above)
Comment (ktro2828) [reply perception/image_projection_based_fusion/include/image_projection_based_fusion/utils/geometry.hpp]: updated in https://github.com/autowarefoundation/autoware.universe/pull/4147/commits/9fa0df5259dc534859c4b3beda72ab09640a1acd

File: perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp [hunk 3]
```diff
@@ -177,7 +179,11 @@ void RoiClusterFusionNode::fuseOnSingleImage(
       if (use_iou_y_) {
         iou_y = calcIoUY(cluster_map.second, feature_obj.feature.roi);
       }
-      if (max_iou < iou + iou_x + iou_y) {
+      if (
+        max_iou < iou + iou_x + iou_y &&
+        ((!only_allow_inside_cluster_) ||
+         (only_allow_inside_cluster_ &&
+          is_inside(feature_obj.feature.roi, cluster_map.second, roi_scale_factor_)))) {
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp:186]: How about this?
```suggestion
      const bool passed_inside_cluster_gate =  only_allow_inside_cluster_ &&
          is_inside(feature_obj.feature.roi, cluster_map.second, roi_scale_factor_)
      if (max_iou < iou + iou_x + iou_y && passed_inside_cluster_gate) {
```

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/utils/geometry.hpp (hunk 2 above)
Comment (yukkysaito) [reply perception/image_projection_based_fusion/include/image_projection_based_fusion/utils/geometry.hpp]: LGTM]

File: perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp (hunk 1 above)
Comment (yukkysaito) [reply perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp]: LGTM

File: perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp (hunk 3 above)
Comment (ktro2828) [reply perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp:186]: If those, doesn't `passed_inside_cluster` always return `false` when `only_allow_inside_cluster=false`?

Comment (yukkysaito) [reply perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp:186]: Exactly
How about this?
```suggestion
      const bool passed_inside_cluster_gate = only_allow_inside_cluster_? is_inside(feature_obj.feature.roi, cluster_map.second, roi_scale_factor_) : true;
      if (max_iou < iou + iou_x + iou_y && passed_inside_cluster_gate) {
```

Comment (ktro2828) [reply perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp:186]: Looks good!

**Review by yukkysaito — APPROVED** [review]
LGTM

**Pull request merged**